therapy/etl/chemidplus.py

```diff
@@ -0,0 +1,179 @@
+import logging
+from typing import Dict
+from boto3.dynamodb.table import BatchWriter
+import re
+
+
+logger = logging.getLogger('therapy')
+logger.setLevel(logging.DEBUG)
+
+
+TAGS_REGEX = r' \[.*\]'
+
+
+class ChemIDplus(Base):
+    """Core ChemIDplus ETL class."""
+
+    def __init__(self,
+                 database: Database,
+                 data_path: Path = PROJECT_ROOT / 'data' / 'chemidplus',
+                 src_server: str = 'ftp.nlm.nih.gov',
+                 src_dir_path: str = 'nlmdata/.chemidlease/',
+                 src_fname: str = 'CurrentChemID.xml'):
+        """Initialize class instance.
+
+        :param therapy.database.Database database: database instance to use
+        :param Path data_path: path to chemidplus subdirectory in application
+            data folder
+        :param str src_dir: URL to remote directory containing source file
+        :param str src_fname: name of file as stored in src_dir.
+
+        If the source file is provided locally in the data_path directory,
+        it's unnecessary to provide `src_dir` and `src_fname` args.
+        """
+        self.database = database
+        # self._src_dir = src_dir
+        self._src_server = src_server
+        self._src_dir_path = src_dir_path
+        self._src_fname = src_fname
+        self._added_ids = set()
+        # perform ETL
+        self._extract_data(data_path)
+        self._add_meta()
+        self._transform_data()
+
+    def _download_data(self, data_path: Path):
+        """Download source data from default location."""
+        logger.info('Downloading ChemIDplus data...')
+        outfile_path = data_path / self._src_fname
+        try:
+            with FTP(self._src_server) as ftp:
+                ftp.login()
+                logger.debug('FTP login successful.')
+                ftp.cwd(self._src_dir_path)
+                with open(outfile_path, 'wb') as fp:
+                    ftp.retrbinary(f'RETR {self._src_fname}', fp.write)
+            logger.info('Downloaded ChemIDplus source file.')
+        except TimeoutError:
+            logger.error('Connection to EBI FTP server timed out.')
+        date = ET.parse(outfile_path).getroot().attrib['date']
+        version = date.replace('-', '')
+        outfile_path.rename(data_path / f'chemidplus_{version}.xml')
+        logger.info('Finished downloading ChemIDplus data')
+
+    def _extract_data(self, data_path: Path):
+        """Acquire ChemIDplus dataset.
+
+        :arg pathlib.Path data_path: directory containing source data
+        """
+        data_path.mkdir(exist_ok=True, parents=True)
+        dir_files = list(data_path.iterdir())
+        if len(dir_files) == 0:
+            self._download_data(data_path)
+            dir_files = list(data_path.iterdir())
+        file = sorted([f for f in dir_files
+                       if f.name.startswith('chemidplus')])
+        self._data_src = file[-1]
+        self._version = self._data_src.stem.split('_')[1]
```

Comment:
I think it'd be useful if we added another check if there are files in the dir but not the correct one. 
```suggestion
        dir_files = list(data_path.iterdir())
        if len(dir_files) == 0:
            file = self._get_file(data_path)
        else:
            file = sorted([f for f in dir_files if
                           f.name.startswith('chemidplus')])
            if not file:
                file = self._get_file(data_path)
        self._data_src = file[-1]
        self._version = self._data_src.stem.split('_')[1]

    def _get_file(self, data_path):
        self._download_data(data_path)
        dir_files = list(data_path.iterdir())
        return sorted([f for f in dir_files if
                       f.name.startswith('chemidplus')])
```

Reply:
✅